File: invoice-management-system/process-invoice-emails-cloud-function/process_invoice_emails_cloud_function/vendors/vendors_client.py
import datetime
import requests
from typing import Union
from .vendor import Vendor
from .error_response import ErrorResponse


class VendorsClient:
    def __init__(self, base_url: str):
        self._vendors_service_base_url = base_url

    def list_vendors(self, email=None) -> Union[list[Vendor], ErrorResponse]:
        params = {}

        if email is not None:
            params["email"] = email

        list_vendors_response = requests.get(
            self._vendors_service_base_url,
            params=params,
            headers={"Accept": "application/json"},
        )

        list_vendors_response_json = list_vendors_response.json()

        if list_vendors_response.status_code != 200:
            return ErrorResponse(
                code=list_vendors_response_json["error"]["code"],
                message=list_vendors_response_json["error"]["message"],
            )

        vendors = []

        for vendor in list_vendors_response_json:
            vendors.append(
                Vendor(
                    _id=vendor["id"],
                    name=vendor["name"],
                    email=vendor["email"],
                    created_at=datetime.datetime.fromisoformat(vendor["createdAt"]),
                    updated_at=datetime.datetime.fromisoformat(vendor["updatedAt"]),
                )
            )

        return vendors

File: invoice-management-system/invoices-management-app/src/common/clients/api/vendors/vendors-client.ts
import axios, {isAxiosError} from 'axios';
import {Vendor} from './models';
import {ErrorResponse} from '../errors';

interface VendorsClientOptions {
  baseUrl: string;
}

interface CreateVendorOptions {
  name: string;
  email: string;
}

interface ListVendorsOptions {
  orderBy?: OrderByClause[];
}

interface OrderByClause {
  field: 'name';
  direction: 'asc' | 'desc';
}

class VendorsClient {
  constructor(private readonly options: VendorsClientOptions) {}

  async createVendor(options: CreateVendorOptions): Promise<Vendor> {
    try {
      const {data: vendor} = await axios.post(this.options.baseUrl, options);

      return this.transformVendorResponse(vendor);
    } catch (err) {
      throw this.tryMakeErrorResponse(err);
    }
  }

  async getVendorById(vendorId: string): Promise<Vendor> {
    try {
      const {data: vendor} = await axios.get(
        `${this.options.baseUrl}/${vendorId}`
      );

      return this.transformVendorResponse(vendor);
    } catch (err) {
      throw this.tryMakeErrorResponse(err);
    }
  }

  async listVendors(options?: ListVendorsOptions): Promise<Vendor[]> {
    try {
      const params: {orderBy?: string[]} = {};

      if (options?.orderBy) {
        params.orderBy = options.orderBy.map(
          this.orderByClauseToQueryParamClause
        );
      }

      const {data: vendors} = await axios.get(this.options.baseUrl, {
        params,
      });

      return vendors.map(this.transformVendorResponse);
    } catch (err) {
      throw this.tryMakeErrorResponse(err);
    }
  }

  async deleteVendorById(vendorId: string): Promise<void> {
    try {
      await axios.delete(`${this.options.baseUrl}/${vendorId}`);
    } catch (err) {
      throw this.tryMakeErrorResponse(err);
    }
  }

  private orderByClauseToQueryParamClause(
    orderByClause: OrderByClause
  ): string {
    return `${orderByClause.field} ${orderByClause.direction}`;
  }

  private transformVendorResponse(vendor: Vendor): Vendor {
    return {
      ...vendor,
      createdAt: new Date(vendor.createdAt),
      updatedAt: new Date(vendor.updatedAt),
    };
  }

  private tryMakeErrorResponse(err: unknown) {
    if (isAxiosError(err)) {
      if (err.response) {
        const errorData = err.response.data['error'];

        return new ErrorResponse(errorData.code, errorData.message);
      }
    }

    return err;
  }
}

export {VendorsClient};
